src/Evaluator.hs

```diff
@@ -38,8 +103,343 @@ data EvalError
   | EvalErrorType TypeError
   | EvalErrorCase SrcLoc Value
   | EvalErrorUser Syntax
-  deriving (Show)
+  | EvalErrorIdent Value
+  deriving (Show, Typeable)
 makePrisms ''EvalError
+instance Exception EvalError
+
+-- | The Kontinuation type. The naming convention InFoo means that the subject
+-- of evaluation in the CEK machine is Foo. For example, when the continuation
+-- is 'InArg' the subject of evaluation is the argument of the function and the
+-- continuation holds the un-evaluated function symbol in its first field.
+data Kont where
+
+  Halt :: Kont
+  -- ^ Marks the evaluator finishing
+
+  -- functions
+  InArg :: !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
+  -- ^ The argument is being evaluated, so hold onto the function symbol.
+  InFun :: !Value -> !VEnv -> !Kont -> Kont
```

Comment:
Why is the `VEnv` needed? It will not be used to further evaluate the `Value`, since it is already a `Value`. Is it perhaps so that the debugger can display the captured environment at this point in the stack trace?

Reply:
Turns out it is needed by the `Er` constructor, in case an error occurs, for example if the function-position value did not evaluate to a Closure.